File: client/request.go
// Copyright 2020 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

package client

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

type ScBaseInfo struct {
	request ScMutableMap
}

func NewScBaseInfo(key MapKey, function string) ScBaseInfo {
	requests := root.GetMapArray(key)
	request := requests.GetMap(requests.Length())
	request.GetString(KeyFunction).SetValue(function)
	return ScBaseInfo{request}
}

func (ctx ScBaseInfo) Contract(contract string) ScBaseInfo {
	ctx.request.GetString(KeyContract).SetValue(contract)
	return ctx
}

func (ctx ScBaseInfo) exec(delay int64) {
	ctx.request.GetInt(KeyDelay).SetValue(delay)
}

func (ctx ScBaseInfo) Params() ScMutableMap {
	return ctx.request.GetMap(KeyParams)
}

func (ctx ScBaseInfo) results() ScImmutableMap {
	return ctx.request.GetMap(KeyResults).Immutable()
}

func (ctx ScBaseInfo) transfer(color *ScColor, amount int64) {
	transfers := ctx.request.GetMap(KeyTransfers)
	transfers.GetInt(color).SetValue(amount)
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

type ScCallInfo struct {
	ScBaseInfo
}

func (ctx ScCallInfo) Call() ScCallInfo {
	ctx.exec(-1)
	return ctx
}

func (ctx ScCallInfo) Results() ScImmutableMap {
	return ctx.ScBaseInfo.results()
}

func (ctx ScCallInfo) Transfer(color *ScColor, amount int64) {
	ctx.ScBaseInfo.transfer(color, amount)
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

type ScPostInfo struct {
	ScBaseInfo
}

func (ctx ScPostInfo) Chain(chain *ScAddress) ScPostInfo {
	ctx.request.GetAddress(KeyChain).SetValue(chain)
	return ctx
}

func (ctx ScPostInfo) Post(delay int64) {
	ctx.exec(delay)
}

func (ctx ScPostInfo) Transfer(color *ScColor, amount int64) {
	ctx.ScBaseInfo.transfer(color, amount)
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

type ScViewInfo struct {
	ScBaseInfo
}

func (ctx ScViewInfo) Results() ScImmutableMap {
	return ctx.ScBaseInfo.results()
}

func (ctx ScViewInfo) View() ScViewInfo {
	ctx.exec(-2)
	return ctx
}

File: rust/wasplib/src/client/request.rs
// Copyright 2020 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

use super::hashtypes::*;
use super::immutable::*;
use super::keys::*;
use super::mutable::*;

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

pub struct ScBaseInfo {
    request: ScMutableMap,
}

impl ScBaseInfo {
    fn new<T: MapKey + ?Sized>(key: &T, function: &str) -> ScBaseInfo {
        let requests = ROOT.get_map_array(key);
        let request = requests.get_map(requests.length());
        request.get_string(&KEY_FUNCTION).set_value(function);
        ScBaseInfo { request: request }
    }

    fn chain(&self, chain: &ScAddress) {
        self.request.get_address(&KEY_CHAIN).set_value(chain);
    }

    fn contract(&self, contract: &str) {
        self.request.get_string(&KEY_CONTRACT).set_value(contract);
    }

    fn exec(&self, delay: i64) {
        self.request.get_int(&KEY_DELAY).set_value(delay);
    }

    fn params(&self) -> ScMutableMap {
        self.request.get_map(&KEY_PARAMS)
    }

    fn results(&self) -> ScImmutableMap {
        self.request.get_map(&KEY_RESULTS).immutable()
    }

    fn transfer(&self, color: &ScColor, amount: i64) {
        let transfers = self.request.get_map(&KEY_TRANSFERS);
        transfers.get_int(color).set_value(amount);
    }
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

pub struct ScCallInfo {
    base: ScBaseInfo,
}

impl ScCallInfo {
    pub fn new(function: &str) -> ScCallInfo {
        ScCallInfo { base: ScBaseInfo::new(&KEY_CALLS, function) }
    }

    pub fn call(&self) {
        self.base.exec(-1);
    }

    pub fn contract(&self, contract: &str) -> &ScCallInfo {
        self.base.contract(contract);
        self
    }

    pub fn params(&self) -> ScMutableMap {
        self.base.params()
    }

    pub fn results(&self) -> ScImmutableMap {
        self.base.results()
    }

    pub fn transfer(&self, color: &ScColor, amount: i64) {
        self.base.transfer(color, amount);
    }
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

pub struct ScPostInfo {
    base: ScBaseInfo,
}

impl ScPostInfo {
    pub fn new(function: &str) -> ScPostInfo {
        ScPostInfo { base: ScBaseInfo::new(&KEY_POSTS, function) }
    }

    pub fn chain(&self, chain: &ScAddress) -> &ScPostInfo {
        self.base.chain(chain);
        self
    }

    pub fn contract(&self, contract: &str) -> &ScPostInfo {
        self.base.contract(contract);
        self
    }

    pub fn params(&self) -> ScMutableMap {
        self.base.params()
    }

    pub fn post(&self, delay: i64) {
        self.base.exec(delay);
    }

    pub fn transfer(&self, color: &ScColor, amount: i64) {
        self.base.transfer(color, amount);
    }
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

pub struct ScViewInfo {
    base: ScBaseInfo,
}

impl ScViewInfo {
    pub fn new(function: &str) -> ScViewInfo {
        ScViewInfo { base: ScBaseInfo::new(&KEY_VIEWS, function) }
    }

    pub fn contract(&self, contract: &str) -> &ScViewInfo {
        self.base.contract(contract);
        self
    }

    pub fn params(&self) -> ScMutableMap {
        self.base.params()
    }

    pub fn results(&self) -> ScImmutableMap {
        self.base.results()
    }

    pub fn view(&self) {
        self.base.exec(-2);
    }
}
